### PR TITLE
Adding fuller support for resolving anonymous type data on a dynamic object

### PR DESCRIPTION
**Pull request by holdenmai**
Issue #93 

The major change made with this is that rather than performing the binding, somewhat greedily, on typeof(object), we leverage the actual instance present in the CallSiteBinder args parameter to get the actual type as late as possible.

### CODE REVIEW

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 1]
```diff
@@ -1676,43 +1677,38 @@ private Expression ParseNormalMethodInvocation(Type type, Expression instance, i
...
 		{
-			var binder = Microsoft.CSharp.RuntimeBinder.Binder.GetMember(
-				Microsoft.CSharp.RuntimeBinder.CSharpBinderFlags.None,
-				propertyOrFieldName,
-				type,
-				new[] { Microsoft.CSharp.RuntimeBinder.CSharpArgumentInfo.Create(Microsoft.CSharp.RuntimeBinder.CSharpArgumentInfoFlags.None, null) }
-			);
+			if (t == null || t.IsArray)
+			{
+				return null;
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:1690]: If it works with null, it doesn't really matter, but maybe the binder requires the array's element type?

```c#
return t.GetElementType();
```

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 2]
```diff
@@ -3340,5 +3336,73 @@ private static T[] RemoveLast<T>(T[] array)
 				return result;
 			}
 		}
+
+		/// <summary>
+		/// Binds to a member access of an instance as late as possible.  This allows the use of anonymous types on dynamic values.
+		/// </summary>
+		private class LateGetMemberCallSiteBinder : CallSiteBinder
+		{
+			private readonly string m_propertyOrFieldName;
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:3345]: To be consistent with naming conventions, please use `_propertyOrFieldName`

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 3]
```diff
@@ -3340,5 +3336,73 @@ private static T[] RemoveLast<T>(T[] array)
...
+			{
+				var binder =
+					Microsoft.CSharp.RuntimeBinder.Binder.GetMember(
+					Microsoft.CSharp.RuntimeBinder.CSharpBinderFlags.None,
+					m_propertyOrFieldName,
+					RemoveArrayType(args[0]?.GetType()),
+					new[] { Microsoft.CSharp.RuntimeBinder.CSharpArgumentInfo.Create(Microsoft.CSharp.RuntimeBinder.CSharpArgumentInfoFlags.None, null) }
+				);
+				var e = binder.Bind(args, parameters, returnLabel);
+				return e;
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:3362]: `return  binder.Bind(args, parameters, returnLabel);`

File: src/DynamicExpresso.Core/Parsing/Parser.cs [hunk 4]
```diff
@@ -3340,5 +3336,73 @@ private static T[] RemoveLast<T>(T[] array)
...
+
+			public LateGetMemberCallSiteBinder(string propertyOrFieldName)
+			{
+				m_propertyOrFieldName = propertyOrFieldName;
+			}
+
+			public override Expression Bind(object[] args, ReadOnlyCollection<ParameterExpression> parameters, LabelTarget returnLabel)
+			{
+				var binder =
+					Microsoft.CSharp.RuntimeBinder.Binder.GetMember(
```
Comment (metoule) [inline src/DynamicExpresso.Core/Parsing/Parser.cs:3355]: [NIT] `var binder = Microsoft.CSharp.RuntimeBinder.Binder.GetMember(` on the same line

**Review by metoule — COMMENTED** [review]
I have a few minor nitpicking comments, but otherwise it looks good to me, thanks a lot!

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 1 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs:1690]: It seems like it works just fine whether we return null or the element type.  This might be a detail that will require a very specific edge case for someone to encounter that isn't already covered by the unit tests to figure out which one is needed.

File: test/DynamicExpresso.UnitTest/DynamicTest.cs [hunk 5]
```diff
@@ -20,7 +20,20 @@ public void Get_Property_of_an_ExpandoObject()
 			var interpreter = new Interpreter()
 				.SetVariable("dyn", (object)dyn);
 
-			Assert.AreEqual(dyn.Foo, interpreter.Eval("dyn.Foo"));
+			var l = interpreter.Parse("dyn.Foo");
+			l.Invoke();
+			l.Invoke();
+
+			//Assert.AreEqual(dyn.Foo, interpreter.Eval("dyn.Foo"));
```
[the remark this answers is not in the document]
Comment (holdenmai) [reply test/DynamicExpresso.UnitTest/DynamicTest.cs]: Reverted/resolved in latest commit

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 2 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: Resolved in latest commit

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 3 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: Resolved in latest commit

File: src/DynamicExpresso.Core/Parsing/Parser.cs (hunk 4 above)
Comment (holdenmai) [reply src/DynamicExpresso.Core/Parsing/Parser.cs]: Resolved in latest commit

**Review by davideicardi — APPROVED** [review]
LGTM! Very nice improvements!

Thank you @holdenmai !
